tool/tctl/common/bots_command.go

```diff
@@ -238,9 +283,156 @@ func (c *BotsCommand) AddBot(ctx context.Context, client auth.ClientI) error {
+		if err := client.UpsertToken(ctx, token); err != nil {
+			return trace.Wrap(err)
+		}
+	} else {
+		// If there is, check the token matches the potential bot
+		token, err = client.GetToken(ctx, c.tokenID)
+		if err != nil {
+			if trace.IsNotFound(err) {
+				return trace.NotFound("token with name %q not found, create the token or do not set TokenName: %v",
+					c.tokenID, err)
+			}
+			return trace.Wrap(err)
+		}
+		if !token.GetRoles().Include(types.RoleBot) {
+			return trace.BadParameter("token %q is not valid for role %q",
+				c.tokenID, types.RoleBot)
+		}
+		if token.GetBotName() != c.botName {
+			return trace.BadParameter("token %q is valid for bot with name %q, not %q",
+				c.tokenID, token.GetBotName(), c.botName)
+		}
+	}
+
+	bot := &machineidv1pb.Bot{
+		Metadata: &headerv1.Metadata{
+			Name: c.botName,
+		},
+		Spec: &machineidv1pb.BotSpec{
+			Roles: roles,
+			Traits: []*machineidv1pb.Trait{
+				{
+					Name:   constants.TraitLogins,
+					Values: flattenSlice(c.allowedLogins),
+				},
+			},
+		},
+	}
+
+	bot, err = client.BotServiceClient().CreateBot(ctx, &machineidv1pb.CreateBotRequest{
+		Bot: bot,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if c.format == teleport.JSON {
+		tokenTTL := time.Duration(0)
+		if exp := token.Expiry(); !exp.IsZero() {
+			tokenTTL = time.Until(exp)
+		}
+		// This struct is equivalent to a legacy bit of JSON we used to output
+		// when we called an older RPC. We've preserved it here to avoid
+		// breaking customer scripts.
+		response := struct {
+			UserName string        `json:"user_name"`
+			RoleName string        `json:"role_name"`
+			TokenID  string        `json:"token_id"`
+			TokenTTL time.Duration `json:"token_ttl"`
+		}{
+			UserName: bot.Status.UserName,
+			RoleName: bot.Status.RoleName,
+			TokenID:  token.GetName(),
+			TokenTTL: tokenTTL,
+		}
+		out, err := json.MarshalIndent(response, "", "  ")
+		if err != nil {
+			return trace.Wrap(err, "failed to marshal CreateBot response")
+		}
+
+		fmt.Println(string(out))
+		return nil
+	}
+
+	proxies, err := client.GetProxies()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(proxies) == 0 {
+		return trace.Errorf("This cluster does not have any proxy servers running.")
+	}
+	addr := proxies[0].GetPublicAddr()
+	if addr == "" {
+		addr = proxies[0].GetAddr()
+	}
+
+	joinMethod := token.GetJoinMethod()
+	if joinMethod == types.JoinMethodUnspecified {
+		joinMethod = types.JoinMethodToken
+	}
+
+	return startMessageTemplate.Execute(os.Stdout, map[string]interface{}{
+		"token":       token.GetName(),
+		"minutes":     int(time.Until(token.Expiry()).Minutes()),
+		"addr":        addr,
+		"join_method": joinMethod,
+	})
+}
+
 func (c *BotsCommand) RemoveBot(ctx context.Context, client auth.ClientI) error {
-	if err := client.DeleteBot(ctx, c.botName); err != nil {
-		return trace.WrapWithMessage(err, "error deleting bot")
+	_, err := client.BotServiceClient().DeleteBot(ctx, &machineidv1pb.DeleteBotRequest{
```

Comment:
Could use a similar note about checking for versioning as `AddBot` has.

Reply:
Yup - I'll do this tomorrow.

Reply:
e472f48327